pyquil/api/_qam.py

```diff
@@ -60,18 +61,27 @@ def load(self, executable):
         return self
 
     @_record_call
-    def write_memory(self, *, region_name: str, offset: int = 0, value=None):
+    def write_memory(self, *, region_name: str, offset: Optional[int] = None,
+                     value: Optional[Union[int, float, Sequence[int], Sequence[float]]] = None):
         """
-        Writes a value into a memory region on the QAM at a specified offset.
+        Writes a value or chronologically unwraps a list of values into a memory region on
+        the QAM at a specified offset.
 
         :param region_name: Name of the declared memory region on the QAM.
         :param offset: Integer offset into the memory region to write to.
-        :param value: Value to store at the indicated location.
+        :param value: Value(s) to store at the indicated location.
         """
         assert self.status in ['loaded', 'done']
 
-        aref = ParameterAref(name=region_name, index=offset)
-        self._variables_shim[aref] = value
+        if isinstance(value, Sequence):
+            if offset is not None:
+                warnings.warn("offset should be None when value is a Sequence")
+            for index, v in enumerate(value):
+                aref = ParameterAref(name=region_name, index=offset + index)
```

Comment:
`offset + index` will raise an exception when `offset is None`. likewise in the `else` clause below, passing `index=offset` is probably not valid when `offset is None`.

I would move this warning and the `offset is None` check and warning to the outer scope, and set a default value of 0 if `offset is None` (and additionally warn if `value` is a `Sequence`).

Reply:
for clarity, something like this moved to the outer scope:

```python
if offset is None:
    offset = 0
elif isinstance(value, Sequence):
    warnings.warn("blahaa")
```